### PR TITLE
Bring your own vNet for the application gateway and upgrade Open Liberty Operator to 0.8.2

### LINKED ISSUE #47: Support bringing your own vNet for Application Gateway
The issue is to request add support for bringing user's own existing vNet or create a new vNet with user inputs for Azure Application Gateway in Liberty on AKS solution offer. See the user experience design in the section within the red-line box from the screen shot below:

image

### LINKED ISSUE #49: Upgrade Open Liberty Operator to 0.8.2
Upgrade Open Liberty Operator to 0.8.2, see [installation guide](https://github.com/OpenLiberty/open-liberty-operator/tree/main/deploy/releases/0.8.2#installation).

### PR DESCRIPTION
**Pull request by majguo**
## Description

The PR is to resolve the following issues:
* Resolve #47
* Resolve #49

## Change summary

* Add a new section **Configure virtual networks** into blade `Networking`.
* Add new Bicep modules for deploying Azure virtual network and related Azure resources.
* Update scripts to configure network peering between Application Gateway vNet and AKS vNet.
* Add Bicep module and scripts to query the private ip to support assigning private ip to the frontend of Application Gateway. 
  * **Note**: currently the feature of assigning a private ip to application gateway is hidden from the UI due to a known issue. It won't be exposed unless the issue is resolved, and all stakeholders agree to open the feature switch.
* Upgrade Open Liberty Operator to 0.8.2

## Testing

The following test cases have already been passed before opening the PR:

* Successfully deployed the offer with new vNet for Application Gateway + deploy user own app
* Successfully deployed the offer with an existing vNet for Application Gateway + deploy user own app

Here is the [preview link](https://portal.azure.com/?feature.customportal=false#create/microsoft_javaeeonazure_test.20191212-arm-rhel-was-nd-v9-cluster-previewazure-liberty-aks) for live testing.

## UI changes review

Besides visiting [preview link](https://portal.azure.com/?feature.customportal=false#create/microsoft_javaeeonazure_test.20191212-arm-rhel-was-nd-v9-cluster-previewazure-liberty-aks) for reviewing new blade `Networking`, reviewer can preview the UI changes using the **Create UI Definition Sandbox**:

- Open [Create UI Definition Sandbox](https://portal.azure.com/?feature.customPortal=false#blade/Microsoft_Azure_CreateUIDef/SandboxBlade) in a new tab of browser.
- Open UI source code [createUiDefinition.txt](https://raw.githubusercontent.com/majguo/azure.liberty.aks/bring-own-vnet-appgw/src/main/arm/createUiDefinition.json) in a new tab of the browser.
- Replace the content of work area in **Create UI Definition Sandbox** with the content of **UI source code `createUiDefinition.json`**.
- Click **Preview** at the left-bottom of **Create UI Definition Sandbox**.
- Switch to `Networking` tab > Choose `Yes` for `Connect to Azure Application Gateway?` > Locate to section `Configure virtual networks` > Start UI content review. **Note**: **Configure virtual networks** is a standard component from Azure Portal, only the following text can be customized:
   - Label `Virtual network` and its tooltip `Select a virtual network in which to place the Application Gateway.`
   - Label `Subnet` and its tooltip `The subnet must be dedicated only for use by the Application Gateway.`

Here is the screenshot of user experience design, see **Configure virtual networks** within the red-line box added to blade `Networking`:
image

Signed-off-by: Jianguo Ma <jiangma@microsoft.com>

### CODE REVIEW

File: src/main/scripts/createAppGatewayIngress.sh [hunk 1]
```diff
@@ -52,31 +52,52 @@ function network_peers_aks_appgw() {
...
-    utility_validate_status "Create network peers for $aksNetWorkId and ${appgwVNetName}."
+    appGatewaySubnetId=$(az network application-gateway show -g ${curRGName} --name ${appgwName} -o tsv --query "gatewayIpConfigurations[0].subnet.id")
+    appGatewayVnetResourceGroup=$(az network application-gateway show -g ${curRGName} --name ${appgwName} -o tsv --query "gatewayIpConfigurations[0].subnet.resourceGroup")
+    appGatewaySubnetName=$(az resource show --ids ${appGatewaySubnetId} --query "name" -o tsv)
+    appgwNetworkId=$(echo $appGatewaySubnetId | sed s/"\/subnets\/${appGatewaySubnetName}"//)
+    appgwVnetName=$(az resource show --ids ${appgwNetworkId} --query "name" -o tsv)
+
+    local toPeer=true
+    # if the AKS and App Gateway have the same VNET, need not peer.
+    if [ "${aksNetWorkId}" == "${appgwNetworkId}" ]; then
```
Comment (edburns) [inline src/main/scripts/createAppGatewayIngress.sh:63]: I have asked this general question in several code reviews: why does this script not use the important best practices for shell scripting, such as those in [https://transang.me/best-practice-to-make-a-shell-script/](https://transang.me/best-practice-to-make-a-shell-script/)? 

In particular 

```
#!/usr/bin/env bash
set -Eeuo pipefail
```

If we had that, I would worry less about the many invocations of sub commands such as the `az network vnet` one below and elsewhere.

Comment (edburns) [reply src/main/scripts/createAppGatewayIngress.sh:63]: I will accept this now, but I would like to understand why we don't need the best practices here.

File: src/main/scripts/install.sh [hunk 2]
```diff
@@ -100,6 +100,44 @@ wait_service_available() {
     done
 }
 
+wait_ingress_available() {
```
Comment (edburns) [inline src/main/scripts/install.sh:103]: Same story here.  Why no

```
#!/usr/bin/env bash
set -Eeuo pipefail
```

File: src/main/scripts/queryPrivateIPForAppGateway.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
```
Comment (edburns) [inline src/main/scripts/queryPrivateIPForAppGateway.sh:1]: Same story here.  Why no

```
#!/usr/bin/env bash
set -Eeuo pipefail
```

**Review by edburns — APPROVED** [review]

File: src/main/scripts/createAppGatewayIngress.sh (hunk 1 above)
Comment (majguo) [reply src/main/scripts/createAppGatewayIngress.sh:63]: Good point. Here is some of background why some of these flags are not applied. For example, we want to output explicit and meaningful error message if script failed, which is helpful for user to trouble shoot the issue. Another example is that some of command invoked in the script may return non-zero value, but it doesn't mean it's a fatal error that the execution should exit. 

I do agree applying best practice of shell programming is meaningful, so https://github.com/WASdev/azure.liberty.aks/issues/50 is created for tracking the requirement of enhancement and let's move forward as for now.

File: src/main/scripts/install.sh (hunk 2 above)
Comment (majguo) [reply src/main/scripts/install.sh:103]: See [response ](https://github.com/WASdev/azure.liberty.aks/pull/48#discussion_r919625547)above.

File: src/main/scripts/queryPrivateIPForAppGateway.sh (hunk 3 above)
Comment (majguo) [reply src/main/scripts/queryPrivateIPForAppGateway.sh:1]: See [response ](https://github.com/WASdev/azure.liberty.aks/pull/48#discussion_r919625547)above.

**Review by git4rk — APPROVED** [review]

**Pull request merged**